### PR TITLE
[fix] Added error handling in task.perform_change_of_authorization

### PR DESCRIPTION
**Pull request by pandafy**
## Checklist

- [x] I have read the [OpenWISP Contributing Guidelines](http://openwisp.io/docs/developer/contributing.html).
- [ ] I have manually tested the changes proposed in this pull request.
- [x] I have written new test cases for new code and/or updated existing tests for changes to existing code.
- [ ] I have updated the documentation.

--- 

How a user can encounter this edge case:

1. A user initially belongs to a RADIUS Group that provides unrestricted access (`power-users`).
2. The user consumes more than 1 GB of traffic.
3. The user’s RADIUS group is then downgraded to a RADIUS Group with a daily limit of 100 MB (Max-Daily-Session-Traffic).
4. When executing DailyTrafficCounter.check(),  the counter will raise MaxQuotaReached because the user has consumed more than 100 MB on the same day.

### CODE REVIEW

File: openwisp_radius/tasks.py [hunk 1]
```diff
@@ -147,15 +148,22 @@ def get_radsecret_from_radacct(rad_acct):
...
+            Counter = app_settings.CHECK_ATTRIBUTE_COUNTERS_MAP[check.attribute]
+            counter = Counter(user=user, group=check.group, group_check=check)
             value = counter.check()
             return counter.reply_name, value
-        except KeyError:
-            return check.attribute, check.value
         except Exception as e:
-            logger.exception(f"Got {e} while CoA for counter {Counter}")
+            # Ignore MaxQuotaReached and KeyError exceptions:
+            # - MaxQuotaReached: raised by Counter.check() when user has
```
Comment (nemesifier) [inline openwisp_radius/tasks.py]: If `MaxQuotaReached` is raised, wouldn't be more correct to `return  counter.reply_name, 0`?

Or am I missing something?

File: openwisp_radius/tasks.py [hunk 2]
```diff
@@ -147,15 +148,21 @@ def get_radsecret_from_radacct(rad_acct):
...
-        counter = Counter(user=user, group=check.group, group_check=check)
         try:
+            Counter = app_settings.CHECK_ATTRIBUTE_COUNTERS_MAP[check.attribute]
+            counter = Counter(user=user, group=check.group, group_check=check)
             value = counter.check()
             return counter.reply_name, value
-        except KeyError:
-            return check.attribute, check.value
+        except MaxQuotaReached:
+            return check.attribute, 0
```
Comment (nemesifier) [inline openwisp_radius/tasks.py:157]: I gave it a deeper though and I am not sure this makese sense. What's the use of this?
It looks like we're just trying to handle the error without performing any useful action.

I think we need to discuss this further.

Can you please write in the PR description what's the possibles scenarios in which this happens?

Comment (pandafy) [reply openwisp_radius/tasks.py:157]: I have update d the PR description.

Comment (nemesifier) [reply openwisp_radius/tasks.py:157]: Thanks @pandafy, so if that's the case, we should log out the user, shouldn't we?

Comment (pandafy) [reply openwisp_radius/tasks.py:157]: In my testing, if we send a CoA packet with limits less than the amount user has consumed (e.g. limit is 100 MB and user has consumed 1 GB), the captive portal automatically logs out the user. 

RFC defines a special kind of CoA packet called ["Disconnect Messages"](https://wiki.freeradius.org/protocol/Disconnect-Messages). This will require adding a new method in the RadClient class. 

https://github.com/openwisp/openwisp-radius/blob/3d6264b31b10e3316de529aecb5f6e876e77e9c3/openwisp_radius/radclient/client.py#L29-L73

Comment (nemesifier) [reply openwisp_radius/tasks.py:157]: What if the max quota is due to time usage? Setting Session-Timeout to 0 AFAIK means disabling it completely. 

For that reason I think sending an explicit disconnect may be more reliable and probably more readable.

**Comment by pandafy** [comment]
Superseded by https://github.com/openwisp/openwisp-radius/pull/645
